rclpy/rclpy/callback_groups.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from threading import Lock
+import weakref
+
+
+class CallbackGroup:
+    """Control when callbacks are allowed to be executed."""
+
+    def __init__(self):
+        super().__init__()
+        self.entities = set()
+
+    def add_entity(self, entity):
+        """
+        Add an entity to the callback group.
+
+        :param entity: a subscription, timer, client, or service instance
+        :rtype: None
+        """
+        self.entities.add(weakref.ref(entity))
+
+    def has_entity(self, entity):
+        """
+        Determine if an entity has been added to this group.
+
+        :param entity: a subscription, timer, client, or service instance
+        :rtype: bool
+        """
+        return weakref.ref(entity) in self.entities
+
+    def can_execute(self, entity):
+        """
+        Return true if an entity can be executed.
+
+        :param entity: a subscription, timer, client, or service instance
+        :rtype: bool
+        """
+        raise NotImplementedError
```

Comment:
`NotImplementedError` is an exception class and should therefore be instantiated: `raise NotImplementedError()`.

Same below.

Reply:
Exceptions are instantiated in c74a9947f761c9e8aca703e464d185e1ebc50f19

Side note, it should be equivalent because of some sneaky behavior. [PEP 3109](https://www.python.org/dev/peps/pep-3109/) says when a subclass of BaseException is raised it gets instantiated by calling it with no arguments.